Add KPs to retriever

colleenXu opened the pull request:
Discuss before merging! @tokebe @maximusunc 

In this PR, I've adjusted the list of Automat KPs:
* added robokop
* made extra adjustments. If we don't want these, I can revert the most recent commit
    * removed the APIs that robokop covers
    * added KPs that robokop doesn't cover (icees-kg, viral-proteome were kept from the original list in retriever):
        * cam-kp
        * genome-alliance
        * monarch-kg

I had issues running retriever, so I did integration testing with BTE. BTE seemed to work with all the added KPs. 
* I set INSTANCE_ENV=test
* These APIs have `sri_testing_data` endpoints, for example edges...which was awesome for my testing! However, I noticed that I couldn't find some of the example edges that I queried for, @maximusunc.

colleenXu commented:
@tokebe @maximusunc 

I know that we also wanted to add RTX-KG2 and MolePro to retriever. But I encountered some issues...

### RTX-KG2

BTE does seem to work with this KP API. But I saw some odd edges returned - qualifiers for non-chem-gene edges and a qualifier with multiple contradicting values (not what I would expect for an Edge's qualifier set). 

<details><summary>What I did</summary>
<p>

First, I adjusted the API list to only include RTX-KG2 (TRAPI 1.5 registration):

```
  - id: a6b575139cfd429b0a87f825a625d036
    name: RTX KG2 - TRAPI 1.5.0
```

Then I ran a smartapi_sync / ran BTE with INSTANCE_ENV=test. 

<details><summary>I ran this query to lookup what pathways a gene was in</summary>
<p>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": ["NCBIGene:2739"],
                    "categories":["biolink:Gene"]
                },
                "n1": {
                    "categories":["biolink:Pathway"]
                }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1"
                }
            }
        }
    }
}
```

</p>
</details> 

<details><summary>And I got edges like this, where I'm confused by the use of qualifiers in the first place, and the multiple values for `object_direction_qualifier` and `object_aspect_qualifier`. </summary>
<p>

```
                "26b2a93b9a47dea7f9fcc289e9bd99e0": {
                    "predicate": "biolink:related_to",
                    "subject": "NCBIGene:2739",
                    "object": "GO:0006749",
                    "qualifiers": [
                        {
                            "qualifier_type_id": "biolink:qualified_predicate",
                            "qualifier_value": [
                                "biolink:causes"
                            ]
                        },
                        {
                            "qualifier_type_id": "biolink:object_direction_qualifier",
                            "qualifier_value": [
                                "increased",
                                "decreased",
                                "downregulated"
                            ]
                        },
                        {
                            "qualifier_type_id": "biolink:object_aspect_qualifier",
                            "qualifier_value": [
                                "activity_or_abundance",
                                "activity"
                            ]
                        }
                    ],
                    "attributes": [
                        {
                            "attribute_type_id": "biolink:original_predicate",
                            "description": "The IDs of the original RTX-KG2pre edge(s) corresponding to this edge prior to any synonymization or remapping.",
                            "value": [
                                "ENSEMBL:ENSG00000124767---biolink:related_to---None---None---None---GO:0006749---identifiers_org_registry:ensembl"
                            ],
                            "value_type_id": "metatype:String"
                        }
                    ],
                    "sources": [
                        {
                            "resource_id": "infores:ensembl-gene",
                            "resource_role": "primary_knowledge_source"
                        },
                        {
                            "resource_id": "infores:rtx-kg2",
                            "resource_role": "aggregator_knowledge_source",
                            "upstream_resource_ids": [
                                "infores:ensembl-gene"
                            ]
                        },
                        {
                            "resource_id": "infores:biothings-explorer",
                            "resource_role": "aggregator_knowledge_source",
                            "upstream_resource_ids": [
                                "infores:rtx-kg2"
                            ]
                        }
                    ]
                },
```

</p>
</details> 

I wondered if BTE was merging records/edges on accident, but the console logs don't appear to show that any merging happened (only 1 record per edge made):

```
  bte:biothings-explorer-trapi:QueryResult result ID: n0-NCBIGene:2739_&_n1-GO:0006749 has 1 +1ms
  bte:biothings-explorer-trapi:QueryResult result ID: n0-NCBIGene:2739_&_n1-GO:0030316 has 1 +0ms
```

</p>
</details> 



### MolePro

I think BTE has trouble parsing the `/meta_knowledge_graph` contents from the API. 

<details><summary>What I did</summary>
<p>

First, I adjusted the API list to only include MolePro (TRAPI 1.5 registration):

```
  - id: 1901bab8d33bb70b124f400ec1cfdba3
    name: MolePro
```

Then I ran a smartapi_sync / ran BTE with INSTANCE_ENV=test (I also tested with ci and dev, with the same outcome)

<details><summary>I ran this query, which is the example provided in the smartapi registration</summary>
<p>

This is the example from https://smart-api.info/ui/1901bab8d33bb70b124f400ec1cfdba3#/query/post_query

```
{
  "message": {
    "query_graph": {
      "edges": {
        "e00": {
          "subject": "n00",
          "predicates": [
            "biolink:affects"
          ],
          "object": "n01"
        }
      },
      "nodes": {
        "n00": {
          "ids": [
            "PUBCHEM.COMPOUND:2244"
          ],
          "categories": [
            "biolink:SmallMolecule"
          ]
        },
        "n01": {
          "categories": [
            "biolink:Gene"
          ]
        }
      }
    }
  }
}

```

</p>
</details> 

BTE returns 0 results, and the response logs show an odd error regarding the metaKG. 

```
    "logs": [
        {
            "timestamp": "2024-05-24T05:05:08.232Z",
            "level": "INFO",
            "message": "Expanded ids for node n00: (1 ids -> 1 ids)",
            "code": null
        },
        {
            "timestamp": "2024-05-24T05:05:08.966Z",
            "level": "ERROR",
            "message": "Something has gone wrong and the MetaKG is empty.Please try again later.If this persists, please contact the server admin.",
            "code": null
        }
    ]
```

<details><summary>Strangely, the console logs don't show any error</summary>
<p>

```
  bte:biothings-explorer-trapi:main Expanded ids for node n00: (1 ids -> 1 ids) +0ms
  bte:biothings-explorer-trapi:main Start to load metakg. +511ms
  bte:biothings-explorer-trapi:main Query options are: {"apiList":{"include":[{"id":"1901bab8d33bb70b124f400ec1cfdba3","name":"MolePro"}],"exclude":[]},"schema":"3.0.0","provenanceUsesServiceProvider":false,"resolveOutputIDs":true,"metakg":"","smartapi":""} +0ms
  bte:biothings-explorer-trapi:main SmartAPI Specs read from path: /Users/colleenxu/Desktop/biothings_explorer/packages/bte-server/data/smartapi_specs.json +0ms
  bte:smartapi-kg:SyncLoader Using api list spec sync loader now. +0ms
  bte:smartapi-kg:AllSpecsSyncLoader Fetching from file path: /Users/colleenxu/Desktop/biothings_explorer/packages/bte-server/data/smartapi_specs.json +0ms
  bte:smartapi-kg:AllSpecsSyncLoader Hits in inputs: true +60ms
  bte:smartapi-kg:SyncLoader Using all specs sync loader now. +123ms
  bte:smartapi-kg:AllSpecsSyncLoader Fetching from file path: /Users/colleenxu/Desktop/biothings_explorer/packages/bte-server/data/smartapi_specs.json +63ms
  bte:smartapi-kg:AllSpecsSyncLoader Hits in inputs: true +65ms
```

</p>
</details> 

</p>
</details>

maximusunc commented:
As an ideally fully-federated system, I strongly advocate for including all KPs, regardless of overlap or unusual (but valid) responses.

As far as with Molepro, I'm not entirely sure what went on there with their /meta_knowledge_graph, but I ran Strider's kp registration service this morning and I didn't see any issues and Molepro was successfully registered.

tokebe commented:
I'll look into what's going wrong with Molepro. As for fully federating, I think that should wait until after the demo. I'm generally in support but there's probably a little more discussion to be had, and room for better implementation in the python migration than in the current iteration of retriever.

tokebe commented:
@colleenXu Update: the reason MolePro isn't working is because every edge in their metaKG has an empty array under `knowledge_types`. BTE/Retriever requires that `knowledge_types` includes `lookup` to be able to add the edge to its predicates data. We can change this to require that the array either includes `lookup` or is empty, [though according to the TRAPI specification, this is invalid](https://github.com/NCATSTranslator/ReasonerAPI/blob/6198a711762ce5615c4bf1a9f5f1f24782ca6635/TranslatorReasonerAPI.yaml#L1367-L1376) (`knowledge_types` must be `null` to imply `lookup`, not an empty array):

colleenXu commented:
@tokebe 

Update on RTX-KG2: the odd qualifiers I mentioned [above](https://github.com/BioPack-team/retriever/pull/2#issuecomment-2128632862) aren't from the original RTX-KG2 response. So I think something is happening on BTE/Retriever's end...(I'm reminded of this [issue](https://github.com/biothings/biothings_explorer/issues/713))

An aside: there's also more edges/results in the direct RTX-KG2 response - that don't show up in BTE's response when hooked up to RTX-KG2. Maybe it has to do with different Pathway namespaces or node/entity resolution? 

<details><summary>More Info</summary>
<p>

I sent the same gene->pathway [above](https://github.com/BioPack-team/retriever/pull/2#issuecomment-2128632862) to RTX-KG2's test instance `https://kg2.test.transltr.io/api/rtxkg2/v1.4/query` and got [this response](https://github.com/BioPack-team/retriever/files/15441672/rtx-kg2.json). 

I found the edge corresponding to my example [above](https://github.com/BioPack-team/retriever/pull/2#issuecomment-2128632862), and it doesn't have any qualifiers:

```
                "30311060": {
                    "attributes": [
                        {
                            "attribute_type_id": "biolink:original_predicate",
                            "description": "The IDs of the original RTX-KG2pre edge(s) corresponding to this edge prior to any synonymization or remapping.",
                            "value": [
                                "ENSEMBL:ENSG00000124767---biolink:related_to---None---None---None---GO:0006749---identifiers_org_registry:ensembl"
                            ],
                            "value_type_id": "metatype:String"
                        }
                    ],
                    "object": "GO:0006749",
                    "predicate": "biolink:related_to",
                    "sources": [
                        {
                            "resource_id": "infores:ensembl-gene",
                            "resource_role": "primary_knowledge_source"
                        },
                        {
                            "resource_id": "infores:rtx-kg2",
                            "resource_role": "aggregator_knowledge_source",
                            "upstream_resource_ids": [
                                "infores:ensembl-gene"
                            ]
                        }
                    ],
                    "subject": "NCBIGene:2739"
                },
```

</p>
</details>

colleenXu commented:
@tokebe @andrewsu 

As for MolePro...maybe we should update Retriever to handle the empty array case? 

I currently see that [MolePro CI](https://molepro-trapi.ci.transltr.io/molepro/trapi/v1.5/meta_knowledge_graph) has updated to report knowledge_types: ["lookup"]. So I tested it with Retriever and the same query as [above](https://github.com/BioPack-team/retriever/pull/2#issuecomment-2128632862) - and things seem to work!

But [Dev](https://translator.broadinstitute.org/molepro/trapi/v1.5/meta_knowledge_graph) and [Test](https://molepro-trapi.test.transltr.io/molepro/trapi/v1.5/meta_knowledge_graph) haven't been updated yet.

colleenXu commented:
@tokebe @maximusunc (3/3)

Umm...so do we want to keep the current Automat KP list (where some have been removed due to overlap with robokop)? Or put the removed Automat KPs back in? 

It sounds like there wasn't a consensus...

colleenXu commented:
@tokebe @maximusunc 

RE: MolePro, I've added it to the API list in anticipation of Jackson's changes. 

RE: Automat, I had a discussion with @andrewsu, and he said that Max's view has been consistent and we should follow it for now. So I've added all TRAPI 1.5 Automat KPs to the API list: 
* the ones I previously removed
* other automat apis I could find, that weren't in BTE's list
